torch/_dynamo/repro/after_aot.py

```diff
@@ -162,6 +180,173 @@ def deferred_for_real_inputs(real_inputs):
     return debug_wrapper
 
 
+# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ #
+#                       REPRO SUPPORT CODE
+# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ #
+
+
+def _stride_or_default(stride, *, shape):
+    return stride if stride is not None else utils.make_contiguous_strides_for(shape)
+
+
+def _dtype_or_default(dtype):
+    return dtype if dtype is not None else torch.float32
+
+
+def _device_or_default(device):
+    return device if device is not None else torch.device("cpu")
+
+
+def _storage_offset_or_default(storage_offset):
+    return storage_offset if storage_offset is not None else 0
+
+
+# TODO: Support bundling the entire repro into a zip file for ease of
+# transferring around
+class InputReader:
+    def __init__(self, save_dir=None):
+        # If None, we will generate random data instead
+        if save_dir is None:
+            log.warning("no save_dir specified, will generate random data")
+        self.store = ContentStoreReader(save_dir) if save_dir is not None else None
+
+    def storage(self, storage_hash, nbytes, *, device=None, dtype_hint=None):
+        device = _device_or_default(device)
+        dtype_hint = _dtype_or_default(dtype_hint)
+        if self.store is not None and storage_hash is not None:
+            try:
+                storage = self.store.read_storage(storage_hash)
+            except FileNotFoundError:
+                pass
+            else:
+                if device != storage.device:
+                    log.warning("device mismatch: %s != %s", device, storage.device)
+                    # TODO: transfer it to the right device?  But failing this
+                    # way would be very mysterious!  Would have been better
+                    # not to store device in the serialized format...
```

Comment:
Feels like an assert?

Reply:
since the point of this is repro generation, it seems best to try our best to get a working repro, and only fail out if there is absolutely nothing else we can do